File: test/pageobjects/CountrySelector.js
import { Key } from 'selenium-webdriver'
import { assert } from 'chai'
import BasePage from './BasePage.js'
import { verifyElementCopy } from '../utils/mochaw'

class CountrySelector extends BasePage {
  async selectorLabel() {
    return this.$('.onfido-sdk-ui-CountrySelector-label')
  }

  async fallbackMessageText() {
    return this.$('.onfido-sdk-ui-CountrySelector-fallbackText')
  }

  async helpIcon() {
    return this.$('.onfido-sdk-ui-CountrySelector-helpIcon')
  }

  async errorIcon() {
    return this.$('.onfido-sdk-ui-CountrySelector-errorIcon')
  }

  async countryFinderInput() {
    return this.$('[data-onfido-qa="countrySelector"] #country-search')
  }

  async countryFinderFirstResult() {
    return this.$(
      '[data-onfido-qa="countrySelector"] #country-search__option--0'
    )
  }

  async countryFinderNoResults() {
    return this.$(
      '[data-onfido-qa="countrySelector"] #country-search__listbox li'
    )
  }

  async submitDocumentBtn() {
    return this.$('[data-onfido-qa="countrySelectorNextStep"]')
  }

  async containerArea() {
    return this.$('.onfido-sdk-ui-CountrySelector-container')
  }

  async selectSupportedCountry() {
    this.searchFor('france')
    this.selectFirstOptionInDropdownMenu()
  }
  async selectFirstOptionInDropdownMenu() {
    this.driver.sleep(250)
    this.countryFinderInput().sendKeys(Key.DOWN)
    this.countryFinderFirstResult().click()
  }

  async searchFor(searchQuery) {
    this.countryFinderInput().sendKeys(searchQuery)
  }

  async clickSubmitDocumentButton() {
    this.submitDocumentBtn().click()
  }

  async verifyTitle(copy) {
    verifyElementCopy(this.title(), copy.country_select.title)
  }

  async verifySelectorLabel(copy) {
    verifyElementCopy(this.selectorLabel(), copy.country_select.search.label)
  }

  async verifyCountryFinderDisplayed() {
    assert.isTrue(
      this.countryFinderInput().isDisplayed(),
      'Test Failed: Country finder input should be displayed'
    )
  }

  async verifyCountryFinderNoResultsMessage() {
    assert.isTrue(
      this.countryFinderNoResults().isDisplayed(),
      'Test Failed: Fallback help message should be displayed'
    )
  }

  async isErrorMessagePresent() {
    const classes = this.containerArea().getAttribute('class').split(' ')
    return classes.includes('onfido-sdk-ui-CountrySelector-errorContainer')
  }

  async verifyCountryNotFoundErrorMessageDisplayed() {
    assert.isTrue(
      this.fallbackMessageText().isDisplayed(),
      'Test Failed: Fallback error message should be displayed'
    )
    assert.isTrue(
      this.errorIcon().isDisplayed(),
      'Test Failed: Fallback error icon should be displayed'
    )
  }

  async verifySubmitDocumentBtnIsDisabled() {
    assert.isFalse(
      this.submitDocumentBtn().isEnabled(),
      'Test Failed: Submit Document button should be disabled'
    )
  }

  async verifySubmitDocumentBtnIsEnabled() {
    assert.isTrue(
      this.submitDocumentBtn().isEnabled(),
      'Test Failed: Submit Document button should not be disabled'
    )
  }
}

export default CountrySelector

File: test/webtest/src/main/java/com/onfido/qa/websdk/page/CountrySelector.java
package com.onfido.qa.websdk.page;

import com.onfido.qa.webdriver.Driver;
import com.onfido.qa.webdriver.common.Page;
import com.onfido.qa.websdk.util.ByUtil;
import org.openqa.selenium.By;
import org.openqa.selenium.Keys;
import org.openqa.selenium.WebElement;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class CountrySelector extends BasePage {

    public static final String SUPPORTED_COUNTRY = "france";

    private static final Logger log = LoggerFactory.getLogger(CountrySelector.class);

    public static final By BASE = By.cssSelector("[data-onfido-qa=\"countrySelector\"]");
    public static final By SEARCH = By.cssSelector("[data-onfido-qa='countrySelector'] #country-search");
    public static final By SUBMIT_BUTTON = ByUtil.onfidoQa("countrySelectorNextStep");
    public static final By ERROR_MESSAGE = By.cssSelector(".onfido-sdk-ui-CountrySelector-fallbackText");

    public CountrySelector(Driver driver) {
        super(driver);
    }

    public <T extends Page> T selectSupportedCountry(Class<T> next) {
        return select(SUPPORTED_COUNTRY, next);
    }

    public CountrySelector select(String search) {
        searchFor(search);
        return selectFirstOptionInDropdownMenu();
    }

    public <T extends Page> T select(String search, Class<T> next) {
        select(search);
        return submit(next);
    }

    public <T extends Page> T submit(Class<T> next) {
        click(SUBMIT_BUTTON);
        return createComponent(next);
    }

    public CountrySelector searchFor(String search) {
        input(SEARCH, search);
        return this;
    }

    public CountrySelector selectFirstOptionInDropdownMenu() {
        sleep(250);
        driver.actions().sendKeys(Keys.DOWN, Keys.ENTER).perform();

        return this;
    }

    public boolean isSubmitBtnEnabled() {
        return driver.findElement(SUBMIT_BUTTON).isEnabled();
    }

    public boolean countryNotFoundMessageIdDisabled() {

        try {
            driver.waitFor(1).visibility(ERROR_MESSAGE);
        } catch (Exception e) {
            log.error("Waiting for error message", e);
        }

        return driver.isInDom(ERROR_MESSAGE) && driver.findElement(ERROR_MESSAGE).isDisplayed();
    }

    @Override
    protected By pageId() {
        return BASE;
    }

    public WebElement countryFinderInput() {
        return driver.findElement(BASE, By.cssSelector("#country-search"));
    }

    public String selectorLabel() {
        return text(By.cssSelector(".onfido-sdk-ui-CountrySelector-label"));
    }
}
